lib/plugins/inventory.js

```diff
@@ -333,10 +333,63 @@ function inject (bot, { version, hideErrors }) {
     bot._client.write('close_window', {
       windowId: window.id
     })
+    copyInventory(window)
     bot.currentWindow = null
     bot.emit('windowClose', window)
   }
 
+  function copyInventory (window) {
+    const slotOffset = window.inventoryStart - bot.inventory.inventoryStart
+    for (let i = window.inventoryStart; i < window.inventoryEnd; i++) {
+      const item = window.slots[i]
+      const slot = i - slotOffset
+      if (item) {
+        item.slot = slot
+      }
+      bot.inventory.slots[slot] = item
+    }
+  }
+
+  function tradeMatch (limitItem, targetItem) {
+    return targetItem.type === limitItem.type && targetItem.count >= limitItem.count
+  }
+
+  function expectTradeUpdate (window) {
+    const trade = window.selectedTrade
+    const hasItem = !!window.slots[2]
+
+    if (hasItem !== tradeMatch(trade.inputItem1, window.slots[0])) {
+      if (trade.hasItems2) {
+        return hasItem !== tradeMatch(trade.inputItem2, window.slots[1])
+      }
+      return true
+    }
+    return false
+  }
+
+  async function waitForWindowUpdate (window, slot) {
+    if (window.type === 'minecraft:inventory') {
+      if (slot >= 1 && slot <= 4) {
+        await once(bot.inventory, 'updateSlot:0')
```

Comment:
This is insufficient for several reasons, need code close to `expectTradeUpdate`